### ISSUE TITLE
Conv3d with specific kernel size outputs inconsistent results between FP16 and FP32 in V100 GPU

### ISSUE THREAD

**Issue by minzhon** [pytorch/pytorch#40275]
## 🐛 Bug

The Conv3d output inconsistent results between float mode and half mode, if the GPU is V100 and the conv kernel size is (3,1,1) and the input/output channel number are multiple of 8.
I could repro this in PyTorch 1.2.0~PyTorch 1.5.1 with GPU V100-SXM2/PCIE 16GB/32GB

## To Reproduce

repro code
```
def diff_conv_half():
    import torch
    for ch_in in range(2, 256):
        for ch_out in range(2, 256):
            conv = torch.nn.Conv3d(ch_in, ch_out, kernel_size=(3, 1, 1), bias=False, padding=(1, 0, 0)).cuda().half()
            x = torch.rand(1, ch_in, 5, 7, 13).cuda().half()
            y_half = conv(x)

            conv = conv.float()
            x = x.float()
            y_full = conv(x)

            similarity = torch.nn.functional.cosine_similarity(y_half.float().view(1,-1), y_full.view(1,-1)).item()
            if (similarity < 1.0 - 1e-5):
                print("{0}\t{1}\t{2:.4f}".format(ch_in, ch_out, similarity))
```

## Expected behavior

A part of the output looks like below. We can see, if the input channel number and the output channel number are multiple of 8, the cosine similarity of half mode and float mode is not 1.0. 
```
8	8	0.8186
8	16	0.8308
8	24	0.7656
8	32	0.7553
8	40	0.7876
8	48	0.7515
8	56	0.7431
8	64	0.7942
...	...	...
16	8	0.6826
16	16	0.8344
16	24	0.8015
16	32	0.7959
16	40	0.8219
16	48	0.7528
16	56	0.7625
16	64	0.8010
...	...	...
```

## Environment

PyTorch version: 1.5.1
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: Tesla V100-PCIE-32GB
Nvidia driver version: 418.87.00
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.1

Versions of relevant libraries:
[pip] numpy==1.16.4
[pip] torch==1.5.1
[pip] torchvision==0.6.0a0+35d732a
[conda] _pytorch_select           0.2                       gpu_0
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.1.243             h6bb024c_0
[conda] magma-cuda100             2.5.0                         1    pytorch
[conda] mkl                       2019.4                      243
[conda] mkl-include               2019.4                      243
[conda] mkl-service               2.0.2            py36h7b6447c_0
[conda] mkl_fft                   1.0.12           py36ha843d7b_0
[conda] mkl_random                1.0.2            py36hd81dba3_0
[conda] numpy                     1.16.4           py36h7e9f1db_0
[conda] numpy-base                1.16.4           py36hde5b4d6_0
[conda] pytorch                   1.5.1           py3.6_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] torchvision               0.6.1                py36_cu101    pytorch

**Comment by xwang233** [comment]
Thanks for reporting the issue! I see that you are using cudnn 7.4.1. Can you try with the latest version 7.6.5 or 8.0 and see if that fixes the problem?

https://developer.nvidia.com/cudnn

**Comment by minzhon** [comment]
@xwang233 ,I can repro this issue with cudnn 7.6.5. The full env info looks like below:
PyTorch version: 1.5.1
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration:
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB
GPU 4: Tesla V100-SXM2-16GB
GPU 5: Tesla V100-SXM2-16GB
GPU 6: Tesla V100-SXM2-16GB
GPU 7: Tesla V100-SXM2-16GB

Nvidia driver version: 440.64
cuDNN version: **/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5**

Versions of relevant libraries:
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              hfd86e86_1
[conda] mkl                       2019.4                      243
[conda] mkl-service               2.0.2            py37h7b6447c_0
[conda] mkl_fft                   1.0.12           py37ha843d7b_0
[conda] mkl_random                1.0.2            py37hd81dba3_0
[conda] numpy                     1.16.4           py37h7e9f1db_0
[conda] numpy-base                1.16.4           py37hde5b4d6_0
[conda] numpydoc                  0.9.1                      py_0
[conda] **pytorch                   1.5.1           py3.7_cuda10.2.89_cudnn7.6.5_0    pytorch**
[conda] torchvision               0.6.1                py37_cu102    pytorch

**Comment by xwang233** [comment]
Thanks for confirming the issue on cudnn 7.6.5. This forum post might be related https://discuss.pytorch.org/t/in-some-condition-conv3d-outputs-are-very-different-according-to-apex-amp-opt-level-on-v100-gpu/82265

We are aware of this conv3d issue with kernel size (3, 1, 1) for cudnn 7.6.5, and it is fixed in cudnn v8 release. So here are several workarounds you can take

- Rebuild pytorch from source with cudnn v8.
- Use a similar approach as those in the forum post that transforms conv3d with (3,1,1) kernel size to (1,1,3). The forum post mentioned that this approach has ~10% performance drop, but restores the top1 accuracy of their ResNet model. 
- We will apply a patch to pytorch to block the use of cudnn 7.6.5 in such case, for example, conv3d with kernel size (3,1,1). This may be available in a near-future nightly release, but not in the next stable release. 

Let's keep this issue open as a tracker for the patch fix.

**Comment by ShoufaChen** [comment]
Hi, @xwang233 , @minzhon 

Now, is there any convenient workaround for this issue?  Thanks.

**Comment by xwang233** [comment]
Hi @ShoufaChen , if you still encounter this issue, you may have to disable cudnn with

```python
with torch.backends.cudnn.flags(enabled=False):
    # your convolution code here
```

**Comment by ShoufaChen** [comment]
Hi, @xwang233 , Thanks very much for your quick reply.

I'd like to know does latest nightly pytorch solve this problem?
Or can I install pytorch which is built with cudnnv8 with conda (not from source).

**Comment by xwang233** [comment]
There are nightly binaries built with cuda 11.0 and cudnn 8. You can try to install and see if that solves the problem. https://pytorch.org/get-started/locally/

There isn't a fix at pytorch side yet (for cudnn 7.6.5), but cudnn v8 may solve the problem.

**Comment by ShoufaChen** [comment]
Thanks. I'll test on cudnn v8 and report results soon.

**Comment by ShoufaChen** [comment]
Hi, @xwang233 ,

Is it OK if I set `torch.backends.cudnn.benchmark = False` to disable cudnn?
Does it works same like `with torch.backends.cudnn.flags(enabled=False):`

**Comment by xwang233** [comment]
You can use `torch.backends.cudnn.enabled = False` to disable cudnn. It's the same as context manager `with`. It affects all script below, but `with` only affects code in it.

**Comment by ShoufaChen** [comment]
I see. Thank you.

**Comment by ShoufaChen** [comment]
cudnn8.0.5 works fine.
